pkg/coscheduling/coscheduling.go

```diff
@@ -0,0 +1,259 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package coscheduling
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/pkg/api/v1/pod"
+	framework "k8s.io/kubernetes/pkg/scheduler/framework/v1alpha1"
+)
+
+// Coscheduling is a plugin that implements the mechanism of gang scheduling.
+type Coscheduling struct {
+	frameworkHandle framework.FrameworkHandle
+	podLister       corelisters.PodLister
+	// Key is the name of Namespace/PodGroup.
+	podGroupInfos sync.Map
+}
+
+// PodGroupInfo is a wrapper to a PodGroup with additional information.
+// TODO implement a timeout based gc for the PodGroupInfos map
+type PodGroupInfo struct {
+	name string
+	// timestamp stores the timestamp of the initialization time of PodGroup.
+	timestamp time.Time
+}
+
+var _ framework.QueueSortPlugin = &Coscheduling{}
+var _ framework.PreFilterPlugin = &Coscheduling{}
+var _ framework.PermitPlugin = &Coscheduling{}
+var _ framework.UnreservePlugin = &Coscheduling{}
+
+const (
+	// Name is the name of the plugin used in Registry and configurations.
+	Name                 = "coscheduling"
+	PodGroupName         = "pod-group.scheduling.sigs.k8s.io/name"
+	PodGroupMinAvailable = "pod-group.scheduling.sigs.k8s.io/min-available"
+	// TODO make this configurable
+	PermitWaitingTime = 1 * time.Second
+)
+
+// Name returns name of the plugin. It is used in logs, etc.
+func (cs *Coscheduling) Name() string {
+	return Name
+}
+
+// New initializes a new plugin and returns it.
+func New(_ *runtime.Unknown, handle framework.FrameworkHandle) (framework.Plugin, error) {
+	podLister := handle.SharedInformerFactory().Core().V1().Pods().Lister()
+	return &Coscheduling{frameworkHandle: handle,
+		podLister: podLister,
+	}, nil
+}
+
+// Less are used to sort pods in the scheduling queue.
+// 1. compare the priority of pod
+// 2. compare the timestamp of the initialization time of PodGroup
+// 3. compare the key of PodGroup
+func (cs *Coscheduling) Less(podInfo1 *framework.PodInfo, podInfo2 *framework.PodInfo) bool {
+	pod1 := podInfo1.Pod
+	pod2 := podInfo2.Pod
+	priority1 := pod.GetPodPriority(pod1)
+	priority2 := pod.GetPodPriority(pod2)
+
+	if priority1 != priority2 {
+		return priority1 > priority2
+	}
+
+	pgInfo1 := cs.getPodGroupInfo(podInfo1)
+	pgInfo2 := cs.getPodGroupInfo(podInfo2)
+	time1 := pgInfo1.timestamp
+	time2 := pgInfo2.timestamp
+
+	if time1.Equal(time2) && pgInfo1.name != "" && pgInfo2.name != "" {
```

Comment:
nit: can you make this like line 88?

`if !time1.Equal(time2) { return ...}`

Reply:
Done